lib/cloud_controller/membership.rb

```diff
@@ -59,27 +59,27 @@ def authorized_spaces_subquery(roles)
     def member_space_ids(roles)
       space_role_subqueries(roles).reduce do |query, subquery|
         query.union(subquery, from_self: false)
-      end&.select_map(:space_id)
+      end&.select(:space_id)
     end
 
     def member_org_ids(roles)
       org_role_subqueries(roles).reduce do |query, subquery|
         query.union(subquery, from_self: false)
-      end&.select_map(:organization_id)
+      end&.select(:organization_id)
     end
 
     private
 
     def role_applies_to_space?(roles, space_id)
       return false unless space_id && contains_space_role?(roles)
 
-      member_space_ids(roles).include?(space_id)
+      member_space_ids(roles).select_map(:space_id).include?(space_id)
```

Comment:
The check could be moved to the database. The current statement selects all space ids and then checks if the given id is included in the list. This can be changed to something like:
```
member_space_ids(roles).where(space_id: space_id).any?
```

Reply:
Excellent point! That should massively cut down scans of role tables

Reply:
Done this, but refactored slightly so that the extra `where` clause is made in each of the subqueries on the role tables that we're doing a `union` on, as I gather this is faster than putting it at a 'higher' level.